packages/tx/src/transaction.ts

```diff
@@ -21,290 +21,255 @@ const N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46
+    const values = rlp.decode(serialized)
+
+    if (!Array.isArray(values)) {
+      throw new Error('Invalid serialized tx input. Must be array')
+    }
+
+    return this.fromValuesArray(values, common)
+  }
+
+  public static fromValuesArray(values: Buffer[], common?: Common) {
+    if (values.length > 9) {
+      throw new Error('Invalid transaction. More values than expected were received')
+    }
+
+    const [nonce, gasPrice, gasLimit, to, value, data, v, r, s] = values
+
+    return new Transaction(
+      common,
+      new BN(nonce),
+      new BN(gasPrice),
+      new BN(gasLimit),
+      to ? new Address(to) : undefined,
+      new BN(value),
+      data || Buffer.from([]),
+      new BN(v),
+      new BN(r),
+      new BN(s),
+    )
+  }
 
   /**
-   * Creates a new transaction from an object with its fields' values.
-   *
-   * @param data - A transaction can be initialized with its rlp representation, an array containing
-   * the value of its fields in order, or an object containing them by name.
-   *
-   * @param opts - The transaction's options, used to indicate the chain and hardfork the
-   * transactions belongs to.
-   *
-   * @note Transaction objects implement EIP155 by default. To disable it, use the constructor's
-   * second parameter to set a chain and hardfork before EIP155 activation (i.e. before Spurious
-   * Dragon.)
-   *
-   * @example
-   * ```js
-   * const txData = {
-   *   nonce: '0x00',
-   *   gasPrice: '0x09184e72a000',
-   *   gasLimit: '0x2710',
-   *   to: '0x0000000000000000000000000000000000000000',
-   *   value: '0x00',
-   *   data: '0x7f7465737432000000000000000000000000000000000000000000000000000000600057',
-   *   v: '0x1c',
-   *   r: '0x5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
-   *   s: '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
-   * };
-   * const tx = new Transaction(txData);
-   * ```
+   * This constructor takes the values, validates them, assigns them and freezes the object.
+   * Use the public static factory methods to assist in creating a Transaction object from varying data types.
+   * @note Transaction objects implement EIP155 by default. To disable it, pass in an `@ethereumjs/common` object set before EIP155 activation (i.e. before Spurious Dragon.)
    */
   constructor(
-    data: Buffer | PrefixedHexString | BufferLike[] | TxData = {},
-    opts: TransactionOptions = {},
+    common: Common | undefined,
+    nonce: BN,
+    gasPrice: BN,
+    gasLimit: BN,
+    to: Address | undefined,
+    value: BN,
+    data: Buffer,
+    v?: BN,
+    r?: BN,
+    s?: BN,
   ) {
-    // instantiate Common class instance based on passed options
-    if (opts.common) {
-      if (opts.chain || opts.hardfork) {
-        throw new Error(
-          'Instantiation with both opts.common, and opts.chain and opts.hardfork parameter not allowed!',
-        )
+    if (!(data instanceof Buffer)) {
```

Comment:
I'm not sure if I understand why this is here. I mean, why is only `data` validated? I'd be ok with no type being validated at runtime.

Reply:
done in https://github.com/ethereumjs/ethereumjs-vm/pull/812/commits/20fd30a3481cfa7d5af94ef70bacc7b9869ec854